src/smc-webapp/library.cjsx

```diff
@@ -0,0 +1,319 @@
+#
+##############################################################################
+
+$              = window.$
+underscore = _ = require('underscore')
+misc           = require('smc-util/misc')
+misc_page      = require('./misc_page')
+os_path        = require('path')
+
+{React, ReactDOM, Actions, Store, Table, rtypes, rclass, Redux}  = require('./smc-react')
+{Col, Row, Button, ButtonGroup, ButtonToolbar, FormControl, FormGroup, Panel, Input, Well, SplitButton, MenuItem, Alert, ListGroup, ListGroupItem} = require('react-bootstrap')
+{Markdown, Space, TimeAgo, ErrorDisplay, Icon, Loading, TimeAgo, Tip, Space} = require('./r_misc')
+{webapp_client} = require('./webapp_client')
+{COLORS}        = require('smc-util/theme')
+
+# src: where the library files are
+# start: open this file after copying the directory
+exports.LIBRARY = LIBRARY =
+    first_steps :
+        src    : '/ext/library/first-steps/src'
+        start  : 'first-steps.tasks'
+
+# used for some styles
+HEIGHT = '275px'
+
+# https://github.com/sagemathinc/cocalc-examples
+exports.examples_path = ROOT = '/ext/library/cocalc-examples'
+
+sortBy = (key) ->
+    (list) ->
+        _.sortBy(list, (k) -> k[key]?.toLowerCase() ? k)
+
+# This is the main library component. It consists of a "selector" and a preview.
+# Later, we probably want to filter by tags, free-text search, ... but for now we just have the main categories
+exports.Library = rclass ({name}) ->
+    displayName : "Library-#{name}"
+
+    reduxProps :
+        "#{name}" :
+            current_path        : rtypes.string
+            library             : rtypes.immutable.Map
+            library_selected    : rtypes.object
+        projects:
+            project_map         : rtypes.immutable
+
+    propTypes :
+        actions  : rtypes.object.isRequired
+
+    getInitialState: ->
+        lang        : 'python'
+        copy        : false
+        show_thumb  : false
+        sorted_docs : undefined
+        metadata    : undefined
+
+    componentDidMount: ->
+        # TODO this isn't working
+        #@scroll_into_view = _.debounce((=> $(ReactDOM.findDOMNode(@refs.selector_list)).find('.active').scrollintoview()), 50)
+
+    componentDidUpdate: (props, state) ->
+
+    init_state: (props) ->
+        meta = props.library.getIn(['examples'])?.metadata
+        docs = props.library.getIn(['examples'])?.documents
+
+        if docs?
+            # sort by a triplet: idea is to have the docs sorted by their category,
+            # where some categories have weights (e.g. "introduction" comes first, no matter what)
+            sortfn = (doc) -> [
+                meta.categories[doc.category].weight ? 0
+                meta.categories[doc.category].name.toLowerCase()
+                doc.title?.toLowerCase() ? doc.id
+            ]
+            sdocs  = _.sortBy(docs, sortfn)
+            @setState
+                copy        : false
+                sorted_docs : sdocs
+                metadata    : meta
```

Comment:
Having `metadata` accessible from props as well as state is an anti-pattern (mostly egregious in a large component like this) because it violates a "single source of truth". If I don't know that `@state.metadata` mirrors `library.getIn(['examples', 'metadata'])` I may easily update the wrong one.

Reply:
Additionally, it's unclear if this means that metadata is *supposed* to be a copy of `library.getIn(['examples', 'metadata'])` that the component *does* modify locally. So then I have to understand all places `@state.metadata` is used and hunt down `library.getIn(['examples', 'metadata'])`. If the intent is just to mirror for convenience, a comment in `getInitialState` should suffice.